File: 02_画像認識/vehicle_recognition.py
import os
import sys
import logging
from argparse import ArgumentParser

import cv2
import numpy as np
from openvino.inference_engine import IECore


MODEL_XML = '/opt/intel/openvino/deployment_tools/tools/model_downloader/intel/vehicle-attributes-recognition-barrier-0042/FP32/vehicle-attributes-recognition-barrier-0042.xml'
MODEL_BIN = '/opt/intel/openvino/deployment_tools/tools/model_downloader/intel/vehicle-attributes-recognition-barrier-0042/FP32/vehicle-attributes-recognition-barrier-0042.bin'

COLOR_LABEL = ('white', 'gray', 'yellow', 'red', 'green', 'blue', 'black')
TYPE_LABEL = ('car', 'van', 'truck', 'bus')


def build_logger():
    logger = logging.getLogger(__name__)
    logging.basicConfig(format="[ %(levelname)s ] %(message)s", level=logging.DEBUG, stream=sys.stdout)
    return logger


def build_argparser():
    parser = ArgumentParser(description='Recognize vehicle attributes.')
    parser.add_argument('-i', '--inputs', required=True, type=str, nargs='+', help='[Required] path to image files')
    return parser


if __name__ == '__main__':

    # ロガーを生成する
    logger = build_logger()

    # 引数を読み込む
    args = build_argparser().parse_args()

    # モデルをロードする
    ie = IECore()
    net = ie.read_network(model=MODEL_XML, weights=MODEL_BIN)
    logger.info('Successfully loaded model.')

    # バッチサイズを指定する
    net.batch_size = len(args.inputs)
    logger.info('batch size: {}'.format(net.batch_size))

    # 入力データを生成する
    n, c, h, w = net.input_info['input'].input_data.shape
    # n, c, h, w = net.inputs['input'].shape
    images = np.ndarray(shape=(n, c, h, w))
    for i in range(n):
        image = cv2.imread(args.inputs[i])
        image = cv2.resize(image, (w, h))
        image = image.transpose((2, 0, 1))  # HWC から CHW へ変換
        images[i, :, :, :] = image

    # データを入力して推論を実行する
    exec_net = ie.load_network(network=net, device_name='CPU')
    res = exec_net.infer(inputs={'input': images})
    logger.info('Inference completed.')

    # 出力層から結果を得る
    res_color, res_type = res['color'], res['type']
    logger.debug('res_color: {}'.format(res_color))
    logger.debug('res_type: {}'.format(res_type))

    # 判定結果を出力する
    for i in range(len(res_color)):
        logger.info('--------------------------------------')
        logger.info('FILE: {}'.format(args.inputs[i]))
        logger.info('COLOR: {}'.format(COLOR_LABEL[np.argmax(res_color[i])]))
        logger.info('TYPE: {}'.format(TYPE_LABEL[np.argmax(res_type[i])]))
    logger.info('--------------------------------------')

File: 02_画像認識/cpp/vehicle_recognition.cpp
#include <iostream>
#include <string>
#include <array>

#include <opencv2/core.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>
#include <inference_engine.hpp>


std::array<std::string, 7> COLOR_LABEL{"white", "gray", "yellow", "red", "green", "blue", "black"};
std::array<std::string, 4> TYPE_LABEL{"car", "van", "truck", "bus"};


int main(int argc, char** argv) {

    /* Check arguments */
    if (argc != 4) {
        std::cerr << "Usage: ./vehicle_recognition <model.xml> <model.bin> <image>" << std::endl;
        return -1;
    }

    std::string model_xml(argv[1]);
    std::string model_bin(argv[2]);
    std::string image_path(argv[3]);

    /* Load IR model and configure batch size */
    InferenceEngine::Core ie;
    InferenceEngine::CNNNetwork network = ie.ReadNetwork(model_xml, model_bin);
    network.setBatchSize(1);

    /* Configure input blobs */
    InferenceEngine::InputInfo::Ptr input_info = network.getInputsInfo().begin()->second;
    input_info->getPreProcess().setResizeAlgorithm(InferenceEngine::RESIZE_BILINEAR);
    input_info->setLayout(InferenceEngine::Layout::NHWC);
    input_info->setPrecision(InferenceEngine::Precision::U8);
    
    /* Configure output blobs */
    InferenceEngine::DataPtr output_info = network.getOutputsInfo().begin()->second;
    output_info->setPrecision(InferenceEngine::Precision::FP32);

    /* Loading model to device */
    InferenceEngine::ExecutableNetwork executable_network = ie.LoadNetwork(network, "CPU");
    InferenceEngine::InferRequest infer_request = executable_network.CreateInferRequest();

    /* Load input image */
    cv::Mat image = cv::imread(image_path);
    if (image.empty()) {
        std::cerr << "Faild to load image file: " << image_path << std::endl;
        return -1;
    }

    /* Set input image to input blob */
    const unsigned int channels = image.channels();
    const unsigned int height = image.size().height;
    const unsigned int width = image.size().width;
    InferenceEngine::TensorDesc tDesc(InferenceEngine::Precision::U8,
                                    {1, channels, height, width},
                                    InferenceEngine::Layout::NHWC);
    InferenceEngine::Blob::Ptr imgBlob = InferenceEngine::make_shared_blob<uint8_t>(tDesc, image.data);
    infer_request.SetBlob("input", imgBlob);

    /* Execute inference */
    infer_request.Infer();

    /* Read outputs */
    InferenceEngine::Blob::Ptr color = infer_request.GetBlob("color");
    InferenceEngine::Blob::Ptr type = infer_request.GetBlob("type");
    auto result_color = color->buffer().as<InferenceEngine::PrecisionTrait<InferenceEngine::Precision::FP32>::value_type*>();
    auto result_type = type->buffer().as<InferenceEngine::PrecisionTrait<InferenceEngine::Precision::FP32>::value_type*>();

    /* Dump outputs */
    std::cout << "Input image path: " << image_path << std::endl;
    std::cout << "    COLOR:" << std::endl;
    for (int i = 0, size = COLOR_LABEL.size(); i < size; i++) {
        std::cout << "        " << COLOR_LABEL[i] << " -> " << result_color[i] << std::endl;
    }
    std::cout << "    TYPE:" << std::endl;
    for (int i = 0, size = TYPE_LABEL.size(); i < size; i++) {
        std::cout << "        " << TYPE_LABEL[i] << " -> " << result_type[i] << std::endl;
    }

    return 0;
}
